# **Wine Quality Prediction using Support Vector Machine text**

# **Objective**

### This project aims to predict the quality of wine based on various physichemical properties usins Support vector Machine(SVM).The aim is to leverage various machine learning algorithms to analyze and predict the quality of wine, facilitating better decision-making for winemakers, sellers, and consumers.

**These datasets include features such as:**

Fixed acidity

Volatile acidity

Citric acid

Residual sugar

Chlorides

Free sulfur dioxide

Total sulfur dioxide

Density

pH

Sulphates

Alcohol

Quality (score between 0 and 10)

# **Import Library**

In [ ]:
import pandas as pd
import numpy as np

# **Import CSV as Dataframe**

In [ ]:
df=pd.read_csv(r'https://raw.githubusercontent.com/Lorddhaval/Dataset/main/WhiteWineQuality.csv', sep=';')

# **Get the first five rows of Dataframe**

In [ ]:
df.head()

fixed acidity  volatile acidity  citric acid  residual sugar  chlorides  \
0            7.0              0.27         0.36            20.7      0.045   
1            6.3              0.30         0.34             1.6      0.049   
2            8.1              0.28         0.40             6.9      0.050   
3            7.2              0.23         0.32             8.5      0.058   
4            7.2              0.23         0.32             8.5      0.058   

   free sulfur dioxide  total sulfur dioxide  density    pH  sulphates  \
0                 45.0                 170.0   1.0010  3.00       0.45   
1                 14.0                 132.0   0.9940  3.30       0.49   
2                 30.0                  97.0   0.9951  3.26       0.44   
3                 47.0                 186.0   0.9956  3.19       0.40   
4                 47.0                 186.0   0.9956  3.19       0.40   

   alcohol  quality  
0      8.8        6  
1      9.5        6  
2     10.1        6  
3      9.9        6  
4      9.9        6

# **Get information of Dataframe**

In [ ]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4898 entries, 0 to 4897
Data columns (total 12 columns):
 #   Column                Non-Null Count  Dtype  
---  ------                --------------  -----  
 0   fixed acidity         4898 non-null   float64
 1   volatile acidity      4898 non-null   float64
 2   citric acid           4898 non-null   float64
 3   residual sugar        4898 non-null   float64
 4   chlorides             4898 non-null   float64
 5   free sulfur dioxide   4898 non-null   float64
 6   total sulfur dioxide  4898 non-null   float64
 7   density               4898 non-null   float64
 8   pH                    4898 non-null   float64
 9   sulphates             4898 non-null   float64
 10  alcohol               4898 non-null   float64
 11  quality               4898 non-null   int64  
dtypes: float64(11), int64(1)
memory usage: 459.3 KB


# **Get the Summary Statistics**

In [ ]:
df.describe()

fixed acidity  volatile acidity  citric acid  residual sugar  \
count    4898.000000       4898.000000  4898.000000     4898.000000   
mean        6.854788          0.278241     0.334192        6.391415   
std         0.843868          0.100795     0.121020        5.072058   
min         3.800000          0.080000     0.000000        0.600000   
25%         6.300000          0.210000     0.270000        1.700000   
50%         6.800000          0.260000     0.320000        5.200000   
75%         7.300000          0.320000     0.390000        9.900000   
max        14.200000          1.100000     1.660000       65.800000   

         chlorides  free sulfur dioxide  total sulfur dioxide      density  \
count  4898.000000          4898.000000           4898.000000  4898.000000   
mean      0.045772            35.308085            138.360657     0.994027   
std       0.021848            17.007137             42.498065     0.002991   
min       0.009000             2.000000              9.000000     0.987110   
25%       0.036000            23.000000            108.000000     0.991723   
50%       0.043000            34.000000            134.000000     0.993740   
75%       0.050000            46.000000            167.000000     0.996100   
max       0.346000           289.000000            440.000000     1.038980   

                pH    sulphates      alcohol      quality  
count  4898.000000  4898.000000  4898.000000  4898.000000  
mean      3.188267     0.489847    10.514267     5.877909  
std       0.151001     0.114126     1.230621     0.885639  
min       2.720000     0.220000     8.000000     3.000000  
25%       3.090000     0.410000     9.500000     5.000000  
50%       3.180000     0.470000    10.400000     6.000000  
75%       3.280000     0.550000    11.400000     6.000000  
max       3.820000     1.080000    14.200000     9.000000

# **Get Column Names**

In [ ]:
df.columns

Index(['fixed acidity', 'volatile acidity', 'citric acid', 'residual sugar',
       'chlorides', 'free sulfur dioxide', 'total sulfur dioxide', 'density',
       'pH', 'sulphates', 'alcohol', 'quality'],
      dtype='object')

# **Get shape of Dataframe**

In [ ]:
df.shape

(4898, 12)

# **Get Unique values(Class or Label) in Y variable**

In [ ]:
df['quality'].value_counts()

quality
6    2198
5    1457
7     880
8     175
4     163
3      20
9       5
Name: count, dtype: int64

In [ ]:
df.groupby('quality').mean()

fixed acidity  volatile acidity  citric acid  residual sugar  \
quality                                                                 
3             7.600000          0.333250     0.336000        6.392500   
4             7.129448          0.381227     0.304233        4.628221   
5             6.933974          0.302011     0.337653        7.334969   
6             6.837671          0.260564     0.338025        6.441606   
7             6.734716          0.262767     0.325625        5.186477   
8             6.657143          0.277400     0.326514        5.671429   
9             7.420000          0.298000     0.386000        4.120000   

         chlorides  free sulfur dioxide  total sulfur dioxide   density  \
quality                                                                   
3         0.054300            53.325000            170.600000  0.994884   
4         0.050098            23.358896            125.279141  0.994277   
5         0.051546            36.432052            150.904598  0.995263   
6         0.045217            35.650591            137.047316  0.993961   
7         0.038191            34.125568            125.114773  0.992452   
8         0.038314            36.720000            126.165714  0.992236   
9         0.027400            33.400000            116.000000  0.991460   

               pH  sulphates    alcohol  
quality                                  
3        3.187500   0.474500  10.345000  
4        3.182883   0.476135  10.152454  
5        3.168833   0.482203   9.808840  
6        3.188599   0.491106  10.575372  
7        3.213898   0.503102  11.367936  
8        3.218686   0.486229  11.636000  
9        3.308000   0.466000  12.180000

# **Define Target (Y) and features (X)**

In [113]:
y = df['quality']

In [114]:
y.shape

(4898,)

In [115]:
y

0       6
1       6
2       6
3       6
4       6
       ..
4893    6
4894    5
4895    6
4896    7
4897    6
Name: quality, Length: 4898, dtype: int64

In [116]:
x = df[['fixed acidity', 'volatile acidity', 'citric acid', 'residual sugar',
        'chlorides', 'free sulfur dioxide', 'total sulfur dioxide', 'density',
        'pH', 'sulphates', 'alcohol']]


In [117]:
x = df.drop(['quality'], axis=1)

In [118]:
x.shape

(4898, 11)

In [119]:
x

fixed acidity  volatile acidity  citric acid  residual sugar  chlorides  \
0               7.0              0.27         0.36            20.7      0.045   
1               6.3              0.30         0.34             1.6      0.049   
2               8.1              0.28         0.40             6.9      0.050   
3               7.2              0.23         0.32             8.5      0.058   
4               7.2              0.23         0.32             8.5      0.058   
...             ...               ...          ...             ...        ...   
4893            6.2              0.21         0.29             1.6      0.039   
4894            6.6              0.32         0.36             8.0      0.047   
4895            6.5              0.24         0.19             1.2      0.041   
4896            5.5              0.29         0.30             1.1      0.022   
4897            6.0              0.21         0.38             0.8      0.020   

      free sulfur dioxide  total sulfur dioxide  density    pH  sulphates  \
0                    45.0                 170.0  1.00100  3.00       0.45   
1                    14.0                 132.0  0.99400  3.30       0.49   
2                    30.0                  97.0  0.99510  3.26       0.44   
3                    47.0                 186.0  0.99560  3.19       0.40   
4                    47.0                 186.0  0.99560  3.19       0.40   
...                   ...                   ...      ...   ...        ...   
4893                 24.0                  92.0  0.99114  3.27       0.50   
4894                 57.0                 168.0  0.99490  3.15       0.46   
4895                 30.0                 111.0  0.99254  2.99       0.46   
4896                 20.0                 110.0  0.98869  3.34       0.38   
4897                 22.0                  98.0  0.98941  3.26       0.32   

      alcohol  
0         8.8  
1         9.5  
2        10.1  
3         9.9  
4         9.9  
...       ...  
4893     11.2  
4894      9.6  
4895      9.4  
4896     12.8  
4897     11.8  

[4898 rows x 11 columns]

# **Get Train Test Split**

In [120]:
from sklearn.model_selection import train_test_split

In [134]:
x_train,x_test,y_train,Y_test=train_test_split(x,y,test_size=0.8,stratify = y,random_state=2529)

In [136]:
x_train.shape,x_test.shape,y_train.shape,Y_test.shape

((979, 11), (3919, 11), (979,), (3919,))

# **Select Model**

In [128]:
from sklearn.svm import SVC


In [129]:
model=SVC()

# **Get Model Train**

In [137]:
model.fit(x_train, y_train)


SVC()

# **Get Model Prediction**

In [138]:
y_pred=model.predict(x_test)

In [139]:
x_test

fixed acidity  volatile acidity  citric acid  residual sugar  chlorides  \
2368            7.4              0.34         0.30           14.90      0.037   
3509            6.4              0.26         0.25           10.70      0.046   
3742            7.0              0.15         0.28           14.70      0.051   
69              7.4              0.24         0.29           10.10      0.050   
928             6.5              0.25         0.35           12.00      0.055   
...             ...               ...          ...             ...        ...   
4618            6.1              0.44         0.28            4.25      0.032   
2249            6.3              0.27         0.18            7.70      0.048   
3179            8.5              0.20         0.40            1.10      0.046   
3387            7.6              0.36         0.49           11.30      0.046   
4416            5.7              0.24         0.47            6.30      0.069   

      free sulfur dioxide  total sulfur dioxide  density    pH  sulphates  \
2368                 70.0                 169.0  0.99698  3.25       0.37   
3509                 66.0                 179.0  0.99606  3.17       0.55   
3742                 29.0                 149.0  0.99792  2.96       0.39   
69                   21.0                 105.0  0.99620  3.13       0.35   
928                  47.0                 179.0  0.99800  3.58       0.47   
...                   ...                   ...      ...   ...        ...   
4618                 43.0                 132.0  0.99160  3.26       0.47   
2249                 45.0                 186.0  0.99620  3.23       0.47   
3179                 31.0                 106.0  0.99194  3.00       0.35   
3387                 87.0                 221.0  0.99840  3.01       0.43   
4416                 35.0                 182.0  0.99391  3.11       0.46   

      alcohol  
2368    10.40  
3509     9.90  
3742     9.00  
69       9.50  
928     10.00  
...       ...  
4618    11.30  
2249     9.00  
3179    10.50  
3387     9.20  
4416     9.75  

[3919 rows x 11 columns]

In [140]:
y_pred.shape

(3919,)

In [141]:
y_pred

array([6, 6, 6, ..., 6, 5, 6])

# **Get Model Evaluation**

In [146]:
from sklearn.metrics import confusion_matrix, classification_report


In [149]:
print(confusion_matrix(Y_test, y_pred))


[[   0    0    5   11    0    0    0]
 [   0    0    4  126    0    0    0]
 [   0    0   75 1091    0    0    0]
 [   0    0   76 1683    0    0    0]
 [   0    0   10  694    0    0    0]
 [   0    0    0  140    0    0    0]
 [   0    0    0    4    0    0    0]]


In [151]:
print(classification_report(Y_test, y_pred))


              precision    recall  f1-score   support

           3       0.00      0.00      0.00        16
           4       0.00      0.00      0.00       130
           5       0.44      0.06      0.11      1166
           6       0.45      0.96      0.61      1759
           7       0.00      0.00      0.00       704
           8       0.00      0.00      0.00       140
           9       0.00      0.00      0.00         4

    accuracy                           0.45      3919
   macro avg       0.13      0.15      0.10      3919
weighted avg       0.33      0.45      0.31      3919



/usr/local/lib/python3.10/dist-packages/sklearn/metrics/_classification.py:1471: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))
/usr/local/lib/python3.10/dist-packages/sklearn/metrics/_classification.py:1471: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))
/usr/local/lib/python3.10/dist-packages/sklearn/metrics/_classification.py:1471: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))


# **Accuracy**

In [152]:
from sklearn.metrics import accuracy_score

In [153]:
accuracy_score(Y_test,y_pred)

0.4485838224036744

# **Explaination**

Interpreting the Accuracy of 0.44
An accuracy of 0.44 (44%) indicates that the model's predictions are correct 44% of the time. This performance is relatively low, suggesting that the model might not be capturing the patterns in the data effectively. Possible reasons for low accuracy include:

**Feature Engineering:** The features may not be informative enough or might need more preprocessing.

**Model Choice:** The model used might not be suitable for the problem.

**Data Quality:** There might be issues with data quality or imbalance.

**Overfitting/Underfitting:** The model might be overfitting to the training data or underfitting the training data.


## **Improvement Strategies**
**Feature Engineering:** Try creating new features or transforming existing ones to better capture the underlying patterns.

**Model Tuning:** Experiment with different models and hyperparameters.

**Cross-Validation:** Use cross-validation to better evaluate model performance.

**Ensemble Methods:** Combine multiple models to improve performance.

Data Augmentation: Augment the dataset or balance it if there is an imbalance in the target classes.

# **Conclusion**
Achieving higher accuracy involves iterating over the steps of data preprocessing, model selection, and evaluation. Continuous refinement of the approach can lead to better performance and more accurate predictions.